modules/sfp_spyonweb.py

```diff
@@ -0,0 +1,367 @@
+# Purpose:      SpiderFoot plug-in to search SpyOnWeb for hosts sharing the
+#               same IP address, Google Analytics code, or Google Adsense code.
+#
+# Author:      Brendan Coles <bcoles@gmail.com>
+#
+# Created:     2018-10-25
+# Copyright:   (c) Brendan Coles 2018
+# Licence:     GPL
+# -------------------------------------------------------------------------------
+
+import json
+import re
+import time
+import datetime
+import socket
+from sflib import SpiderFoot, SpiderFootPlugin, SpiderFootEvent
+
+class sfp_spyonweb(SpiderFootPlugin):
+    """SpyOnWeb:Footprint,Investigate,Passive:Passive DNS:apikey:Search SpyOnWeb for hosts sharing the same IP address, Google Analytics code, or Google Adsense code."""
+
+    # Default options
+    opts = {
+        'api_key': '',
+        'limit': 100,
+        'timeout': 30,
+        'maxage': 1095,   # 3 years
+        'verify': True,
+        'cohostsamedomain': False,
+        'maxcohost': 100
+    }
+
+    # Option descriptions
+    optdescs = {
+        'api_key': "SpyOnWeb API key.",
+        'limit': "Maximum number of results to fetch.",
+        'timeout': "Query timeout, in seconds.",
+        'maxage': "The maximum age of the data returned, in days, in order to be considered valid.",
+        'verify': "Verify co-hosts are valid by checking if they still resolve to the shared IP.",
+        'cohostsamedomain': "Treat co-hosted sites on the same target domain as co-hosting?",
+        'maxcohost': "Stop reporting co-hosted sites after this many are found, as it would likely indicate web hosting.",
+    }
+
+    cohostcount = 0
+    results = None
+    errorState = False
+
+    def setup(self, sfc, userOpts=dict()):
+        self.sf = sfc
+        self.__dataSource__ = "SpyOnWeb"
+        self.results = dict()
+        self.cohostcount = 0                                                                                                                                                                                       
+
+        for opt in userOpts.keys():
+            self.opts[opt] = userOpts[opt]
+
+    # Verify a host resolves to an IP
+    def validateIP(self, host, ip):
+        try:
+            addrs = socket.gethostbyname_ex(host)
+        except BaseException as e:
+            self.sf.debug("Unable to resolve " + host + ": " + str(e))
+            return False
+
+        for addr in addrs:
+            if type(addr) == list:
+                for a in addr:
+                    if str(a) == ip:
+                        return True
+            else:
+                if str(addr) == ip:
+                    return True
+        return False
+
+    # What events is this module interested in for input
+    def watchedEvents(self):
+        return ['IP_ADDRESS', 'INTERNET_NAME', 'DOMAIN_NAME', 'WEB_ANALYTICS_ID']
+
+    # What events this module produces
+    def producedEvents(self):
+        return ['CO_HOSTED_SITE', 'INTERNET_NAME', 'AFFILIATE_DOMAIN', 'WEB_ANALYTICS_ID']
```

Comment:
Should we maybe introduce a module that simply extracts analytics IDs from TARGET_WEB_CONTENT events? That way this module can also act on any analytics IDs extracted from web content taken from scanned sites.

Reply:
Yeah I was thinking the same thing. I can't think of any good reason to not implement something like that.